File: Easy/TwoSum.js
//Given an array of integers nums and an integer target, return indices of the two numbers such that they add up to target.

/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var twoSum = function (nums, target) {
  let numsMap = new Map();

  for (i = 0; i < nums.length; i++) {
    let complimentaryNum = target - nums[i];
    if (numsMap.has(complimentaryNum)) {
      return [numsMap.get(complimentaryNum), i];
    }
    numsMap.set(nums[i], i);
  }
};

console.log(twoSum([2, 7, 11, 15], 9)); // [0,1]
console.log(twoSum([3, 2, 4], 6)); // [1,2]

File: Easy/TwoSum.py
# Given an array of integers nums and an integer target, return the indices i and j such that nums[i] + nums[j] == target and i != j.
# You may assume that every input has exactly one pair of indices i and j that satisfy the condition.
# Return the answer with the smaller index first.
# Example 1:
# Input: 
# nums = [3,4,5,6], target = 7
# Output: [0,1]
# Explanation: nums[0] + nums[1] == 7, so we return [0, 1].
# Example 2:
# Input: nums = [4,5,6], target = 10
# Output: [0,2]

class Solution:
    def twoSum(self, nums: list, target: int) -> list:
        differenceDict = {} #Initializing the dictionary

        for i in range(len(nums)): #Looping through nums
            diff = target - nums[i] #Calculating difference
            if diff in differenceDict: #This means we've seen it already in the dictionary
                return [differenceDict[diff], i] #Returning the pair of indecies that equal the target
            else:
                differenceDict[nums[i]] = i #Otherwise, store current number in dictionary with its index
            
            
solution = Solution()
print(solution.twoSum([4, 5, 6], 10))